src/Chapter2.hs

```diff
@@ -728,7 +751,13 @@ value of the element itself
 🕯 HINT: Use combination of 'map' and 'replicate'
 -}
 smartReplicate :: [Int] -> [Int]
-smartReplicate l = error "smartReplicate: Not implemented!"
+smartReplicate [] = []
+smartReplicate (x: xs) = replicate x x ++ smartReplicate xs
+  where
+    replicate :: Int -> Int -> [Int]
+    replicate n x
+      | n < 1 = []
+      | otherwise = x : replicate (n - 1) x
```

Comment:
his is right! 💯 

You can notice that what is happening is you walk through each element of the list, change it to something else (by using `replicate` function) and then concatenate all of the results into one list.

In Haskell, there is the `map` function that does exactly the first part of it (walk through each element of the list, change it to something else ).

After using `map` you will get a list of lists. To flatten such a list, you can use the `concat` function:

```haskell
ghci> concat [[3, 1, 2], [4, 0, 1]]
[3,1,2,4,0,1]
```

This pattern of using `map` and `concat` is very common, so the standard library has even a standard function `concatMap` which is a combination of `map` and `concat` (as the name suggests).

Finally, `replicate` is already a standard function in Haskell :slightly_smiling_face: 

* http://hackage.haskell.org/package/base-4.14.0.0/docs/Data-List.html#v:replicate

Reply:
Thank you for very useful comment. I will revise it using concatMap and pull request again.